File: backend/routers/commissions.py
from fastapi import APIRouter, Depends, Body
from typing import Optional
from core.security import get_current_user, forbid_demo_write
from services.commission_service import commission_service
from models.commission import ManualCommissionIn

router = APIRouter(prefix="/api/commissions", tags=["commissions"])


@router.get("")
async def list_commissions(mandante_id: Optional[str] = None, user=Depends(get_current_user)):
    return await commission_service.list_commissions(user, mandante_id)


@router.get("/bonus-summary")
async def bonus_summary(user=Depends(get_current_user)):
    return await commission_service.bonus_summary(user)


# Percorsi letterali ("/manual", "/manual/{id}") dichiarati PRIMA di
# "/{cid}/status" e "/{cid}" più sotto: non collidono per forma (numero di
# segmenti diverso), ma stare comunque prima segue la stessa convenzione
# già adottata nel resto del router (percorsi specifici prima di quelli
# parametrizzati).
@router.get("/manual")
async def list_manual_commissions(user=Depends(get_current_user)):
    return await commission_service.list_manual_commissions(user)


@router.post("/manual")
async def create_manual_commission(payload: ManualCommissionIn, user=Depends(forbid_demo_write)):
    return await commission_service.create_manual_commission(user, payload.model_dump())


@router.put("/manual/{cid}")
async def update_manual_commission(cid: str, payload: ManualCommissionIn, user=Depends(forbid_demo_write)):
    await commission_service.update_manual_commission(user, cid, payload.model_dump())
    return {"ok": True}


@router.delete("/manual/{cid}")
async def delete_manual_commission(cid: str, user=Depends(forbid_demo_write)):
    await commission_service.delete_manual_commission(user, cid)
    return {"ok": True}


@router.patch("/{cid}/status")
async def update_commission_status(cid: str, payload: dict = Body(...), user=Depends(forbid_demo_write)):
    await commission_service.update_status(user, cid, payload.get("status"))
    return {"ok": True}


@router.delete("/{cid}")
async def delete_commission(cid: str, user=Depends(forbid_demo_write)):
    await commission_service.delete_commission(user, cid)
    return {"ok": True}

File: frontend/src/api/commissions.js
import api from "../api";

// Astrazione per il dominio "provvigioni" — backend/routers/commissions.py.
// Non compreso nell'elenco iniziale di domini, aggiunto per coerenza.

export function listCommissions(params = {}) {
  const qs = new URLSearchParams();
  for (const [k, v] of Object.entries(params)) {
    if (v !== undefined && v !== null && v !== "") qs.append(k, v);
  }
  const query = qs.toString();
  return api.get(`/commissions${query ? `?${query}` : ""}`).then(({ data }) => data);
}

export function getBonusSummary() {
  return api.get("/commissions/bonus-summary").then(({ data }) => data);
}

export function updateCommissionStatus(id, status) {
  return api.patch(`/commissions/${id}/status`, { status }).then(({ data }) => data);
}

export function deleteCommission(id) {
  return api.delete(`/commissions/${id}`).then(({ data }) => data);
}

// --- Provvigioni inserite manualmente (fuori dal flusso ordini) ---

export function listManualCommissions() {
  return api.get("/commissions/manual").then(({ data }) => data);
}

export function createManualCommission(payload) {
  return api.post("/commissions/manual", payload).then(({ data }) => data);
}

export function updateManualCommission(id, payload) {
  return api.put(`/commissions/manual/${id}`, payload).then(({ data }) => data);
}

export function deleteManualCommission(id) {
  return api.delete(`/commissions/manual/${id}`).then(({ data }) => data);
}
